File: LC/python/Intersection of Two Linked Lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    # @param two ListNodes
    # @return the intersected ListNode
    def getIntersectionNode(self, headA, headB):
        
        if headA == None or headB == None:
            return None
                
        m = self.getListlen(headA)
        n = self.getListlen(headB)
        
        pa = headA
        pb = headB
        
        if m>n:
            for i in range(m-n):
                pa = pa.next
        else:
            for i in range(n-m):
                pb = pb.next
        
        while pa and pb:
            if pa!=pb:
                if pa.next != None and pb.next != None:
                    pa = pa.next
                    pb = pb.next
                else:
                    break
            else:
                return pa
            
        return None
        
    def getListlen(self,head):
        size = 0
        p = head
        while p.next:
            size+=1
            p=p.next
        return size

File: LC/java/Intersection of Two Linked Lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode getIntersectionNode(ListNode headA, ListNode headB) {
        ListNode p1,p2;
        int a=1,b=1;
        
        if(headA==null || headB==null){
            return null;
        }
        
        p1 = headA;
        while(p1.next!=null){
            p1 = p1.next;
            a += 1;
        }
        
        p2 = headB;
        while(p2.next!=null){
            p2 = p2.next;
            b += 1;
        }
        
        if(p1!=p2){
            return null;
        }
        
        if(a>b){
            for(int i = 0; i<a-b; i++){
                headA = headA.next;
            }
        }
        
        if(a<b){
            for(int i = 0; i<b-a; i++){
                headB = headB.next;
            }
        }
        
        while(headA!=null){
            if(headA == headB){
                return headA;
            }
            else{
                headA = headA.next;
                headB = headB.next;
            }
        }
        return null;
    }
}